### PR TITLE
Implementing autocompletion for karmadactl

### PR DESCRIPTION
**Pull request by zhzhuang-zju**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Implementing autocompletion for karmadactl, which can save you a lot of typing.

**Which issue(s) this PR fixes**:
Parts of #5477

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`karmadactl`: Implementing autocompletion for karmadactl to save a lot of typing
```

### CODE REVIEW

**Comment by RainbowMango** [comment]
Looks great! Can't wait to try it!
Hi @ahorine , I guess you might like it.

**Comment by zhzhuang-zju** [comment]
> Looks great! Can't wait to try it!
Hi @ahorine , I guess you might like it.

The introduction of `karmadactl completion` this time significantly reduces the amount of typing required during routine operations. Compared to `kubectl completion`, `karmadactl completion` has made extensive adaptations for `karmadactl`. I will put together a guide on how to use `karmadactl completion` to enhance the operational experience. Everyone is welcome to give it a try~ @ahorine @RainbowMango @mohamedawnallah

**Comment by zhzhuang-zju** [comment]
Currently, `karmadactl completion` support output shell completion code for the specified shell:
- zsh
```source <(karmadactl completion zsh)```
- bash
```source <(karmadactl completion bash)```
Theoretically, it should also support `fish` and `powershell`, though this hasn't been verified yet. Verification and implementation can be done in subsequent steps.

Here are the autocompletion capabilities provided by `karmadactl completion`:
- Autocomplete `karmadactl`
By typing `k` and pressing the `tab` key, `karmadactl` can be autocompleted.
```shell
$ k
... karmadactl ... kubectl ...
```
- Autocomplete `karmadactl` subcommands
By typing `karmadactl ` and pressing the `tab` key, the subcommands supported by `karmadactl` can be autocompleted.
```shell
$ karmadactl 
addons         -- Enable or disable a Karmada addon
annotate       -- Update the annotations on a resource
api-resources  -- Print the supported API resources on the server
```
- Autocomplete flags
Currently supports autocompletion for flags `namespace`, `cluster`, `clusters`, `karmada-context`, and `operation-scope`.
The values autocompleted for `operation-scope` vary depending on the command. For example, the `get` command supports `operation-scope` values of `karmada`, `members`, and `all`, so these options will be available for selection during autocompletion. However, when autocompleting the `operation-scope` flag for the `exec` command, only `karmada` and `members` will appear.
```shell
$ karmadactl exec --operation-scope 
karmada  members
$ karmadactl get --operation-scope 
all      karmada  members
```

- Autocomplete resources that can be operated on by commands
For example, for the `get` command:
```shell
$ karmadactl get 
apiservices.apiregistration.k8s.io                            endpoints                                                     nodes.metrics.k8s.io                                          resourceregistries.search.karmada.io                        
...
$ karmadactl get pods 
nginx  pod1 
```

**Highlights**
- The autocompletion content varies depending on the command. For example, the `operation-scope` flag. This not only helps complete the content but also assists users in entering the correct values.
- When autocompleting resources for commands, it takes into account the flags `operation-scope`, `cluster`, and `clusters`.

Using the `get` command as an example:
- `karmadactl get pods` will autocomplete the list of pods in the Karmada control plane.
- `karmadactl get pods --operation-scope all` will autocomplete the list of pods in the Karmada control plane and all member clusters.
- `karmadactl get pods --operation-scope members` will autocomplete the list of pods in all member clusters.
- `karmadactl get pods --operation-scope members --clusters member1,member2` will autocomplete the list of pods in member clusters `member1` and `member2`.

**Comment by Vacant2333** [comment]
its a great improvement for karmadactl, do u need update autocompletion to the karmada docs?

**Comment by zhzhuang-zju** [comment]
> its a great improvement for karmadactl, do u need update autocompletion to the karmada docs?

yes, we can publish a guidance document as long as this pr is merged. Are you willing to submit a pr for this?

File: pkg/karmadactl/completion/completion.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+		    source $HOME/.bash_profile
+
+		# Load the %[1]s completion code for zsh[1] into the current shell
+		    source <(%[1]s completion zsh)
+		# Set the %[1]s completion code for zsh[1] to autoload on startup
+		    %[1]s completion zsh > "${fpath[1]}/%[1]s"`))
+)
+
+var (
+	// TODO: support output shell completion code for more specified shell, like `fish` and `powershell`.
```
Comment (RainbowMango) [inline pkg/karmadactl/completion/completion.go:75]: If this is the plan, please track it somewhere like the issue.

File: pkg/karmadactl/top/top_pods.go [hunk 2]
```diff
@@ -115,13 +115,16 @@ func NewCmdTopPod(f util.Factory, parentCommand string, o *PodOptions, streams g
...
-		ValidArgsFunction:     completion.ResourceNameCompletionFunc(f, "pod"),
+		ValidArgsFunction:     utilcomp.ResourceNameCompletionFunc(f, "pod"),
 		Run: func(cmd *cobra.Command, args []string) {
 			cmdutil.CheckErr(o.Complete(f, cmd, args))
 			cmdutil.CheckErr(o.Validate())
 			cmdutil.CheckErr(o.RunTopPod(f))
 		},
 		Aliases: []string{"pods", "po"},
+		Annotations: map[string]string{
+			"parent": "top",
```
Comment (RainbowMango) [inline pkg/karmadactl/top/top_pods.go:126]:
```suggestion
			"parent": "top", // used for completion code to set default operation scope.
```
I suggest adding a comment for this annotation.

File: pkg/karmadactl/top/top_node.go [hunk 3]
```diff
@@ -104,13 +104,16 @@ func NewCmdTopNode(f util.Factory, parentCommand string, o *NodeOptions, streams
...
-		ValidArgsFunction:     completion.ResourceNameCompletionFunc(f, "node"),
+		ValidArgsFunction:     utilcomp.ResourceNameCompletionFunc(f, "node"),
 		Run: func(cmd *cobra.Command, args []string) {
 			cmdutil.CheckErr(o.Complete(f, cmd, args))
 			cmdutil.CheckErr(o.Validate())
 			cmdutil.CheckErr(o.RunTopNode(f))
 		},
 		Aliases: []string{"nodes", "no"},
+		Annotations: map[string]string{
+			"parent": "top",
```
Comment (RainbowMango) [inline pkg/karmadactl/top/top_node.go:115]:
```suggestion
			"parent": "top",  // used for completion code to set default operation scope.
```

File: pkg/karmadactl/taint/taint.go [hunk 4]
```diff
@@ -77,13 +78,16 @@ var (
...
+	validArgs := []string{"cluster"}
+
 	cmd := &cobra.Command{
 		Use:                   "taint CLUSTER NAME KEY_1=VAL_1:TAINT_EFFECT_1 ... KEY_N=VAL_N:TAINT_EFFECT_N",
 		Short:                 "Update the taints on one or more clusters",
 		Long:                  taintLong,
 		Example:               fmt.Sprintf(taintExample, parentCommand),
 		SilenceUsage:          true,
 		DisableFlagsInUseLine: true,
+		ValidArgsFunction:     utilcomp.SpecifiedResourceTypeAndNameCompletionFunc(f, validArgs),
```
Comment (RainbowMango) [inline pkg/karmadactl/taint/taint.go:90]:
```suggestion
		ValidArgsFunction:     utilcomp.SpecifiedResourceTypeAndNameCompletionFunc(f, []string{"cluster"}),
```

**Review by RainbowMango — COMMENTED** [review]
Just some nits, otherwise LGTM.

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve
Great job!

**Pull request merged**